packages/remix-react/routes.tsx

```diff
@@ -126,27 +125,63 @@ export function createClientRoutes(
   needsRevalidation?: Set<string>
 ): DataRouteObject[] {
   return (routesByParentId[parentId] || []).map((route) => {
+    let routeModule = routeModulesCache?.[route.id];
     let dataRoute: DataRouteObject = {
-      caseSensitive: route.caseSensitive,
-      element: <RemixRoute id={route.id} />,
-      errorElement:
-        route.id === "root" || route.hasErrorBoundary ? (
-          <RemixRouteError id={route.id} />
-        ) : undefined,
       id: route.id,
       index: route.index,
       path: route.path,
-      // handle gets added in via useMatches since we aren't guaranteed to
-      // have the route module available here
-      handle: undefined,
-      loader: createDataFunction(route, routeModulesCache, false),
-      action: createDataFunction(route, routeModulesCache, true),
-      shouldRevalidate: createShouldRevalidate(
-        route,
-        routeModulesCache,
-        needsRevalidation
-      ),
+      loader({ request }) {
+        if (!route.hasLoader) return null;
+        return fetchServerHandler(request, route);
+      },
+      action({ request }) {
+        if (!route.hasAction) {
+          let msg =
+            `Route "${route.id}" does not have an action, but you are trying ` +
+            `to submit to it. To fix this, please add an \`action\` function to the route`;
+          console.error(msg);
+          return Promise.reject(new Error(msg));
+        }
+
+        return fetchServerHandler(request, route);
+      },
```

Comment:
Lifted the forked logic from `createDataFunction` to inline here so we can share 100% of `fetchServerHandler`